extractor/src/main/java/org/schabi/newpipe/extractor/ServiceList.java

```diff
@@ -44,20 +46,33 @@ private ServiceList() {
      * When creating a new service, put this service in the end of this list,
      * and give it the next free id.
      */
-    private static final List<StreamingService> SERVICES = Collections.unmodifiableList(
+    private static List<StreamingService> services = new ArrayList<>(
             Arrays.asList(
                     YouTube = new YoutubeService(0),
                     SoundCloud = new SoundcloudService(1),
                     MediaCCC = new MediaCCCService(2),
                     PeerTube = new PeertubeService(3)
             ));
 
+    public static final int builtinServices = 4;
+    private static int nextService = 4;
+
     /**
      * Get all the supported services.
      *
-     * @return a unmodifiable list of all the supported services
+     * @return an unmodifiable list of all the supported services
      */
     public static List<StreamingService> all() {
-        return SERVICES;
+        return Collections.unmodifiableList(services);
+    }
+
+    public static void addService(final Class<StreamingService> service) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+        services.add(service.getConstructor(new Class[]{int.class}).newInstance(nextService));
+        nextService++;
+    }
+
+    public static void replaceService(final Class<StreamingService> service, final int serviceId) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+        services.set(serviceId, service.getConstructor(new Class[]{int.class}).newInstance(serviceId));
+        nextService++;
```

Comment:
that's a security nightmare

Reply:
Reflection in itself isn't a security nightmare. The NewPipe-side with `PathClassLoader` is perhaps not so great security wise (but other ways of having extensions can't be more secure, except when we add signing), but that's why I've clearly added a message in NewPipe that you should only install extensions that you trust, though perhaps I should emphasize that more.